webview-ui/src/components/chat/ModeSelector.tsx

```diff
@@ -0,0 +1,154 @@
+import { useAppTranslation } from "@/i18n/TranslationContext"
+
+interface ModeSelectorProps {
+	value: Mode
+	onChange: (value: Mode) => void
+	disabled?: boolean
+	title?: string
+	triggerClassName?: string
+	modeShortcutText: string
+	customModes?: any[]
+}
+
+export const ModeSelector: React.FC<ModeSelectorProps> = ({
+	value,
+	onChange,
+	disabled = false,
+	title = "",
+	triggerClassName = "",
+	modeShortcutText,
+	customModes,
+}) => {
+	const [open, setOpen] = React.useState(false)
+	const portalContainer = useRooPortal("roo-portal")
+	const { hasOpenedModeSelector, setHasOpenedModeSelector } = useExtensionState()
+	const { t } = useAppTranslation()
+
+	const trackModeSelectorOpened = () => {
+		if (!hasOpenedModeSelector) {
+			setHasOpenedModeSelector(true)
+			vscode.postMessage({ type: "hasOpenedModeSelector", bool: true })
+		}
+	}
+
+	// Get all available modes
+	const modes = React.useMemo(() => getAllModes(customModes), [customModes])
+
+	// Find the selected mode
+	const selectedMode = React.useMemo(() => modes.find((mode) => mode.slug === value), [modes, value])
+
+	return (
+		<Popover
+			open={open}
+			onOpenChange={(isOpen) => {
+				if (isOpen) trackModeSelectorOpened()
+				setOpen(isOpen)
+			}}
+			data-testid="mode-selector-root">
+			<PopoverTrigger
+				disabled={disabled}
+				title={title}
+				data-testid="mode-selector-trigger"
+				className={cn(
+					"inline-flex items-center gap-1.5 relative whitespace-nowrap px-1.5 py-1 text-xs",
+					"bg-transparent border border-[rgba(255,255,255,0.08)] rounded-md text-vscode-foreground",
+					"transition-all duration-150 focus:outline-none focus-visible:ring-1 focus-visible:ring-vscode-focusBorder focus-visible:ring-inset",
+					disabled
+						? "opacity-50 cursor-not-allowed"
+						: "opacity-90 hover:opacity-100 hover:bg-[rgba(255,255,255,0.03)] hover:border-[rgba(255,255,255,0.15)] cursor-pointer",
+					triggerClassName,
+					!disabled && !hasOpenedModeSelector
+						? "bg-primary opacity-90 hover:bg-primary-hover text-vscode-button-foreground"
+						: null,
+				)}>
+				<ChevronUp className="pointer-events-none opacity-80 flex-shrink-0 size-3" />
+				<span className="truncate">{selectedMode?.name || ""}</span>
+			</PopoverTrigger>
+
+			<PopoverContent
+				align="start"
+				sideOffset={4}
+				container={portalContainer}
+				className="p-0 overflow-hidden w-[320px]">
+				<div className="flex flex-col w-full">
+					<div className="p-3 border-b border-vscode-dropdown-border cursor-default">
+						<div className="flex flex-row items-center gap-1 p-0 mt-0 mb-1 w-full">
+							<h4 className="m-0 pb-2 flex-1">{t("chat:modeSelector.title")}</h4>
+							<div className="flex flex-row gap-1 ml-auto mb-1">
+								<IconButton
+									iconClass="codicon-extensions"
+									title="Mode Marketplace"
+									onClick={() => {
+										vscode.postMessage({
+											type: "switchTab",
+											tab: "marketplace",
+										})
```

Comment:
Would be nice if we could switch directly to the modes tab of the marketplace. We do something similar for links to settings, where we can deep link to a specific settings screen.

Reply:
I built a whole thing for this, extending the ExtensionContext and such, but I wasn't too confident of the implementation. I'll take another stab at it, thanks for the pointer.